crates/sui-core/src/authority/authority_per_epoch_store_util.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::authority_per_epoch_store::DeferralKey;
+use core::cmp::max;
+use narwhal_types::Round;
+use std::collections::HashMap;
+use sui_types::base_types::{ObjectID, TransactionDigest};
+use sui_types::executable_transaction::VerifiedExecutableTransaction;
+use sui_types::transaction::SharedInputObject;
+
+// Given a list of shared input objects, returns the starting cost of a transaction that operates on
+// these objects.
+//
+// Starting cost is a proxy for the starting time of the transaction. It is determined by all the input
+// shared objects' last write.
+pub fn compute_tx_start_at_cost(
+    object_execution_cost: &HashMap<ObjectID, u64>,
+    shared_input_objects: &[SharedInputObject],
+) -> u64 {
+    let mut start_cost = 0;
+    shared_input_objects.iter().for_each(|obj| {
+        let object_cost = *object_execution_cost.get(&obj.id).unwrap_or(&0);
+        start_cost = max(start_cost, object_cost);
+    });
+    start_cost
+}
+
+// Given a transaction, returns the deferral key and the congested objects if the transaction should be deferred.
+pub fn should_defer_due_to_object_congestion(
+    object_execution_cost: &HashMap<ObjectID, u64>,
+    cert: &VerifiedExecutableTransaction,
+    max_accumulated_txn_cost_per_object_in_checkpoint: u64,
+    previously_deferred_tx_digests: &HashMap<TransactionDigest, DeferralKey>,
+    commit_round: Round,
+) -> Option<(DeferralKey, Vec<ObjectID>)> {
+    let shared_input_objects: Vec<_> = cert.shared_input_objects().collect();
+    let start_cost = compute_tx_start_at_cost(object_execution_cost, &shared_input_objects);
+    if start_cost + cert.gas_budget() <= max_accumulated_txn_cost_per_object_in_checkpoint {
+        return None;
+    }
+
+    // Finds out the congested objects.
+    //
+    // Note that the congested objects here may be caused by transaction dependency of other congested objects.
+    // Consider in a consensus commit, there are many transactions touching object A, and later in processing the
+    // consensus commit, there is a transaction touching both object A and B. Although there are fewer transactions
+    // touching object B, becase it's starting execution is delayed due to dependency to other transactions on
+    // object A, it may be shown up as congested objects.
+    let mut congested_objects = vec![];
+    for obj in cert.shared_input_objects() {
+        if &start_cost == object_execution_cost.get(&obj.id).unwrap_or(&0) {
+            congested_objects.push(obj.id);
+        }
+    }
+
+    assert!(!congested_objects.is_empty());
+
+    let deferral_key = if let Some(DeferralKey::ConsensusRound {
+        future_round: _,
+        deferred_from_round,
+    }) = previously_deferred_tx_digests.get(cert.digest())
+    {
+        // This transaction has been deferred before. Use its previous deferred_from_round.
```

Comment:
By "before", do you mean in previous consensus commits? Might need a clarify in the comment